### ISSUE TITLE
Add asynchronous fetch support

### ISSUE THREAD

**Issue by Bilge** [ScriptFUSION/Porter#15]
Performing many sub-imports simultaneously is equivalent to queuing a series of I/O-bound operations whose total execution time is the sum of all imports' individual execution times. By running sub-requests concurrently in parallel asynchronously we reduce the total execution time to that of the the longest-running sub-import only. For highly concurrent sub-imports this is a significant time saving.

**Comment by Bilge** [comment]
The only non-blocking network facility in PHP 7 is the [raw socket](http://php.net/manual/en/function.stream-set-blocking.php).

> This function works for any stream that supports non-blocking mode (currently, regular files and socket streams). 

This has also been independently confirmed by async expert and internals contributor, @bwoebi.

This means that each network protocol has to be implemented using raw socket programming which, needless to say, is extremely laborious and error prone. All the major fledgling async frameworks ([React](http://reactphp.org/), [amphp](http://amphp.org/), [icicle.io](https://icicle.io/)) provide an HTTP client implementation but they have not been battle tested in a wide variety of real-world scenarios like the PHP stream wrapper and may contain bugs. Moreover, they do not provide other protocols found in PHP such as SOAP.

Given the current level of maturity of async frameworks, I do not think it is practical to resolve this issue until async primitives are available in the PHP language or a wider range of async network protocols are supported.

**Comment by Bilge** [comment]
It seems amphp is reaching a reasonable level of maturity, however the latest iteration supports PHP 7 only, so this issue may be on hold until Porter has dropped support for earlier PHP versions.

**Comment by Bilge** [comment]
This has been merged with #54.
